src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java

```diff
@@ -698,4 +699,29 @@ private static void assertRejected(Whitelist whitelist, String expectedSignature
         }
     }
 
+    @Issue("JENKINS-37129")
+    @Test public void methodMissingException() throws Exception {
+        // the case where the unsafe receiver type causes the security check to fail
+        try {
+            assertEvaluate(new GenericWhitelist(), "should fail", "[].noSuchMethod()");
+        } catch (MissingMethodException e) {
+            assertEquals(e.getType(),ArrayList.class);
+            assertThat(e.getMethod(),is("noSuchMethod"));
+        }
+
+        // trying to call an existing method that's not safe
+        try {
+            assertEvaluate(new GenericWhitelist(), "should fail", "[].class.classLoader");
+        } catch (RejectedAccessException e) {
+            assertEquals("method java.lang.Class getClassLoader", e.getSignature());
+        }
+
+        // the case where the safe receiver type causes the security check to pass
```

Comment:
What is the difference between a “safe” and “unsafe” receiver type? I am not following the distinction between the first and third cases here.